[Translation] add types to private properties

rvanlaak opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

Saw @derrabus creating PRs that add types to private properties for other components (#42179, #42109, #42146), so figured out that I could help with the translation component.

Could not find a related ticket with an overview / list of TODOs (?)

Did only add types where certain, there are several places (like `YamlFileLoader`) where the type gets set after the constructor.

rvanlaak committed:
[Translation] add types to private properties

rvanlaak commented:
maybe @Nyholm and @nicolas-grekas can review, as I saw them reviewing other PRs?

stof commented on src/Symfony/Component/Translation/DataCollectorTranslator.php:
```diff
@@ -25,12 +25,9 @@ class DataCollectorTranslator implements TranslatorInterface, TranslatorBagInter
-     * @var TranslatorInterface|TranslatorBagInterface
-     */
-    private $translator;
+    private TranslatorInterface | TranslatorBagInterface $translator;
```
the type should actually be `TranslatorInterface & TranslatorBagInterface & LocaleAwareInterface`, not a union type (see the constructor). So in the meantime, I would use `TranslatorInterface` as the property type (and maybe use phpdoc for the actual intersection type).

Also, the phpdoc should be fixed in older branches to use an intersection type.

rvanlaak replied:
fixed in bd15e17227f4d6847df85b093d55e28f3ddccbf3

stof commented on src/Symfony/Component/Translation/Extractor/PhpExtractor.php:
```diff
@@ -135,7 +133,7 @@ class PhpExtractor extends AbstractFileExtractor implements ExtractorInterface
      * {@inheritdoc}
      */
-    public function extract(string|iterable $resource, MessageCatalogue $catalog)
+    public function extract(string | iterable $resource, MessageCatalogue $catalog)
```
should be reverted (fabbot has a false positive about spaces in union type)

rvanlaak replied:
reverted in f7663551312b74366de018a963b1fd98091631da

stof commented on src/Symfony/Component/Translation/LoggingTranslator.php:
```diff
@@ -21,12 +21,9 @@
-     * @var TranslatorInterface|TranslatorBagInterface
-     */
-    private $translator;
+    private TranslatorInterface | TranslatorBagInterface $translator;
```
Same here. The actual type is an intersection type, not a union type.

rvanlaak replied:
fixed in 5b09467343fe027124d9e7b4f0e150defb7935a2

stof commented on src/Symfony/Component/Translation/MessageCatalogue.php:
```diff
@@ -107,7 +107,7 @@ public function has(string $id, string $domain = 'messages')
         }
 
-        if (null !== $this->fallbackCatalogue) {
+        if (isset($this->fallbackCatalogue)) {
```
no need to change that, as it can indeed be null and no just unset.

rvanlaak replied:
reverted in 301c7ae454beb38c0de75a0d21fba76a3e69efd3

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -37,54 +37,36 @@ class Translator implements TranslatorInterface, TranslatorBagInterface, LocaleA
+    private MessageFormatterInterface $formatter;
 
     /**
      * @var string
```
why keeping a `string` phpdoc if you make it `?string` ?

rvanlaak replied:
fixed in 1516fd39114f2f19942a1ba4cf80ec868d46e181

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -37,54 +37,36 @@ class Translator implements TranslatorInterface, TranslatorBagInterface, LocaleA
-     * @var array|null
-     */
-    private $parentLocales;
+    private ?array $parentLocales;
```
should not be nullable, as it was null before only before initializing it.

rvanlaak replied:
Removed nullable in 1516fd39114f2f19942a1ba4cf80ec868d46e181

Actually, `$parentLocales` does not get initialized in the constructor, so although `computeFallbackLocales` has an `isset` check, are we sure `initializeCatalogue` does not crash because of that?

derrabus replied:
Calling `isset()` on uninitialized properties is safe. Same goes for `??` and `??=`.

derrabus commented on src/Symfony/Component/Translation/Command/XliffLintCommand.php:
```diff
@@ -33,11 +33,13 @@ class XliffLintCommand extends Command
+    /** @var callable|null */
     private $directoryIteratorProvider;
+    /** @var callable|null */
     private $isReadableProvider;
```
```suggestion
    private ?\Closure $directoryIteratorProvider;
    private ?\Closure $isReadableProvider;
```

Would it be possible to make the constructor convert the callables to closures like we did in other PRs?

rvanlaak replied:
Good idea. Fixed in 3cf6702a3dccdcc19ab214869aba2dbeb22cd848

derrabus commented on src/Symfony/Component/Translation/MessageCatalogue.php:
```diff
@@ -19,11 +19,11 @@
+    private array $resources = [];
+    private string $locale;
+    private ?MessageCatalogueInterface $fallbackCatalogue = null;
     private $parent;
```
```suggestion
    private ?self $parent = null;
```

rvanlaak commented:
@stof @derrabus thanks for your reviews. Resolved them, and added the related commits to your comments. Rebased on 6.0 to fix confict. 

Ready for review again. The fabbot check now is a false positive, as @stof mentioned in his review already.

derrabus commented on src/Symfony/Component/Translation/Provider/TranslationProviderCollectionFactory.php:
```diff
@@ -20,8 +20,9 @@
 {
-    private $factories;
-    private $enabledLocales;
+    /** @var ProviderFactoryInterface[] */
```
```suggestion
    /** @var iterable<ProviderFactoryInterface> */
```

rvanlaak replied:
Committed in batch 56b1c3d110f405b279a41d2c64fbbabd3e1629e6

derrabus commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -407,7 +386,7 @@ private function loadFallbackCatalogues(string $locale): void
     protected function computeFallbackLocales(string $locale)
     {
-        if (null === $this->parentLocales) {
+        if (!isset($this->parentLocales)) {
             $this->parentLocales = json_decode(file_get_contents(__DIR__.'/Resources/data/parents.json'), true);
         }
```
```suggestion
        $this->parentLocales ??= json_decode(file_get_contents(__DIR__.'/Resources/data/parents.json'), true);
```

rvanlaak replied:
Committed in batch 56b1c3d110f405b279a41d2c64fbbabd3e1629e6

derrabus commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -464,7 +443,7 @@ protected function assertValidLocale(string $locale)
-        if (!$this->configCacheFactory) {
+        if (!isset($this->configCacheFactory)) {
             $this->configCacheFactory = new ConfigCacheFactory($this->debug);
         }
```
```suggestion
        $this->configCacheFactory ??= new ConfigCacheFactory($this->debug);
```

rvanlaak replied:
Committed in batch 56b1c3d110f405b279a41d2c64fbbabd3e1629e6

derrabus reviewed (changes requested):
Thanks. Just some nit-picking.

rvanlaak commented:
@derrabus as a result of [this recent change](https://github.com/symfony/symfony/commit/8ac2b2a05aeee7c70748c3662a0ac73938ae80b7#diff-aea87a7c5e373feb4fa038e9c8fe1ee760af30fe433b1d09cdb70194d8c48dd4R60) `getOption` would return `null` as default value for the XliffLintCommand's `format` option, resulting in failing tests as private `format` field became a strict string. Committed 13dbc35e8de67826ce5d63cb4972cea18a3af702 to fix that.

Ready for review again 👍

derrabus commented:
> Committed [13dbc35](https://github.com/symfony/symfony/commit/13dbc35e8de67826ce5d63cb4972cea18a3af702) to fix that.

Excellent.

derrabus commented:
Have you checked the Psalm failures?

ro0NL commented on src/Symfony/Component/Translation/Exception/ProviderException.php:
```diff
@@ -20,8 +20,8 @@
-    private $response;
-    private $debug;
+    private ResponseInterface $response;
+    private string $debug;
```
this looks fragile: https://3v4l.org/Epn4h (not sure the concat assignment is intended somehow :/)

derrabus replied:
Good catch. I'd remove the concatenation, tbh. Looks like a mistake.

ro0NL replied:
im curious why psalm doesnt notice :/ https://psalm.dev/r/1f32ea070a

rvanlaak replied:
Given the `\Symfony\Contracts\HttpClient\ResponseInterface::getInfo` docblock can even return an array,  I've written a test to verify that it works in e01134c41840e69045be39f700c711fbd4c76415

Yes, I also think we can see it as typo. [This commit](https://github.com/symfony/symfony/commit/6e55fa84b7ed9751784b4c478d8e16b22062ed93) on 5.3 adds that provider as experimental, but `ProviderException` [is unknown at the `php-translation` organization](https://github.com/search?q=org%3Aphp-translation+ProviderException&type=source) so does not have any other history.

rvanlaak replied:
> im curious why psalm doesnt notice :

Probably because the related error gets included with the baseline errors, as that concat already was part of the codebase before.

> ERROR: UninitializedProperty - 6:9 - Cannot use uninitialized property $this->s

rvanlaak commented:
> Have you checked the Psalm failures?

@derrabus did now with fixes in 6a54b273631792be8a5067a5783af85870980525

but, I'm not sure on how to fix the Psalm issues related to

> [...] with declared type 'Symfony\Contracts\Translation\TranslatorInterface&Symfony\Component\Translation\TranslatorBagInterface&Symfony\Contracts\Translation\LocaleAwareInterface' cannot be assigned type 'Symfony\Component\Translation\TranslatorBagInterface&Symfony\Contracts\Translation\LocaleAwareInterface'

derrabus commented:
@rvanlaak That looks like a false positive to me. Maybe we can fix that by adding the union type to the `@param` annotation of the two constructors as well.

nicolas-grekas commented on src/Symfony/Component/Translation/Command/XliffLintCommand.php:
```diff
@@ -34,18 +34,18 @@ class XliffLintCommand extends Command
-        $this->directoryIteratorProvider = $directoryIteratorProvider;
-        $this->isReadableProvider = $isReadableProvider;
+        $this->directoryIteratorProvider = \is_callable($directoryIteratorProvider) ? \Closure::fromCallable($directoryIteratorProvider) : null;
+        $this->isReadableProvider = \is_callable($isReadableProvider) ? \Closure::fromCallable($isReadableProvider) : null;
```
for both lines:
`$this->foo = null === $foo || $foo instanceof \Closure ? $foo : \Closure::fromCallable($foo)`

nicolas-grekas commented on src/Symfony/Component/Translation/Command/XliffLintCommand.php:
```diff
@@ -84,12 +84,10 @@ protected function execute(InputInterface $input, OutputInterface $output)
-            $this->format = GithubActionReporter::isGithubActionEnvironment() ? 'github' : 'txt';
-        }
+        $format = $input->getOption('format');
+        $this->format = $format ?? GithubActionReporter::isGithubActionEnvironment() ? 'github' : 'txt';
```
`$this->format = $input->getOption('format') ?? GithubActionReporter::isGithubActionEnvironment() ? 'github' : 'txt';`

nicolas-grekas commented on src/Symfony/Component/Translation/DataCollectorTranslator.php:
```diff
@@ -26,11 +26,11 @@ class DataCollectorTranslator implements TranslatorInterface, TranslatorBagInter
 
     /**
-     * @var TranslatorInterface|TranslatorBagInterface
+     * @var TranslatorInterface&TranslatorBagInterface&LocaleAwareInterface
```
can you please send this change in a PR for 5.4?

nicolas-grekas commented on src/Symfony/Component/Translation/DependencyInjection/TranslatorPathsPass.php:
```diff
@@ -53,7 +56,7 @@ public function process(ContainerBuilder $container)
             if ($container->hasDefinition($id)) {
                 [$locatorRef] = $argument->getValues();
-                $this->controllers[(string) $locatorRef][$container->getDefinition($id)->getClass()] = true;
+                $this->controllers[(string) $locatorRef][(string) $container->getDefinition($id)->getClass()] = true;
```
not needed (or for 5.4)

nicolas-grekas commented on src/Symfony/Component/Translation/DependencyInjection/TranslatorPathsPass.php:
```diff
@@ -91,7 +94,7 @@ protected function processValue(mixed $value, bool $isRoot = false)
             if ((string) $value === $this->translatorServiceId) {
                 for ($i = $this->level - 1; $i >= 0; --$i) {
-                    $class = $this->definitions[$i]->getClass();
+                    $class = (string) $this->definitions[$i]->getClass();
```
not needed (or for 5.4)

nicolas-grekas commented on src/Symfony/Component/Translation/DependencyInjection/TranslatorPathsPass.php:
```diff
@@ -22,14 +22,17 @@
+    private array $paths = [];
+    /** @var Definition[] */
+    private array $definitions = [];
+    /** @var array<string, array<string, bool>> with controller as the array key */
```
annotations for 5.4 please

nicolas-grekas commented on src/Symfony/Component/Translation/Exception/ProviderException.php:
```diff
@@ -20,13 +20,13 @@
     {
         $this->response = $response;
-        $this->debug .= $response->getInfo('debug') ?? '';
+        $this->debug = $response->getInfo('debug') ?? '';
```
for 5.4 also

nicolas-grekas commented on src/Symfony/Component/Translation/Extractor/PhpExtractor.php:
```diff
@@ -27,10 +27,8 @@ class PhpExtractor extends AbstractFileExtractor implements ExtractorInterface
     /**
      * Prefix for new found message.
-     *
-     * @var string
```
can be removed on 5.4 too

nicolas-grekas commented on src/Symfony/Component/Translation/LoggingTranslator.php:
```diff
@@ -22,11 +22,11 @@
 {
     /**
-     * @var TranslatorInterface|TranslatorBagInterface
+     * @var TranslatorInterface&TranslatorBagInterface&LocaleAwareInterface
```
for 5.4

nicolas-grekas commented on src/Symfony/Component/Translation/Provider/TranslationProviderCollection.php:
```diff
@@ -20,7 +20,8 @@
 final class TranslationProviderCollection
 {
-    private $providers;
+    /** @var array<string, ProviderInterface> */
```
for 5.4 also

nicolas-grekas commented on src/Symfony/Component/Translation/Provider/TranslationProviderCollectionFactory.php:
```diff
@@ -20,8 +20,9 @@
 {
-    private $factories;
-    private $enabledLocales;
+    /** @var iterable<ProviderFactoryInterface> */
```
for 5.4

nicolas-grekas commented on src/Symfony/Component/Translation/PseudoLocalizationTranslator.php:
```diff
@@ -20,12 +20,13 @@ final class PseudoLocalizationTranslator implements TranslatorInterface
+    private float $expansionFactor;
+    private bool $brackets;
+    private bool $parseHTML;
+    /** @var string[] */
```
for 5.4

nicolas-grekas commented on src/Symfony/Component/Translation/Reader/TranslationReader.php:
```diff
@@ -25,9 +25,9 @@ class TranslationReader implements TranslationReaderInterface
      * Loaders used for import.
      *
-     * @var array
+     * @var array<string, LoaderInterface>
```
for 5.4

nicolas-grekas commented on src/Symfony/Component/Translation/Tests/Command/XliffLintCommandTest.php:
```diff
@@ -167,6 +167,16 @@ public function testLintAutodetectsGithubActionEnvironment()
         }
     }
 
+    public function testPassingClosureAndCallableToConstructor()
```
for 5.4

nicolas-grekas commented on src/Symfony/Component/Translation/Tests/Exception/ProviderExceptionTest.php:
```diff
@@ -0,0 +1,30 @@
+use PHPUnit\Framework\TestCase;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+class ProviderExceptionTest extends TestCase
```
for 5.4

nicolas-grekas commented on src/Symfony/Component/Translation/Writer/TranslationWriter.php:
```diff
@@ -23,7 +23,8 @@
 class TranslationWriter implements TranslationWriterInterface
 {
-    private $dumpers = [];
+    /** @var array<string, DumperInterface> */
```
for 5.4

nicolas-grekas reviewed (commented):
LGTM, but changes that apply to 5.4 should go there.
Can you please send them there in a separate PR?

nicolas-grekas commented:
(please squash also)

derrabus commented:
Backports: #42918

nicolas-grekas commented on src/Symfony/Component/Translation/Command/XliffLintCommand.php:
```diff
@@ -34,18 +34,18 @@ class XliffLintCommand extends Command
-        $this->directoryIteratorProvider = $directoryIteratorProvider;
-        $this->isReadableProvider = $isReadableProvider;
+        $this->directoryIteratorProvider = null === $directoryIteratorProvider || $directoryIteratorProvider instanceof \Closure ? $directoryIteratorProvider : \Closure::fromCallable($directoryIteratorProvider);
+        $this->isReadableProvider = null === $isReadableProvider || $isReadableProvider instanceof \Closure ? null : \Closure::fromCallable($isReadableProvider);
```
s/null/$isReadableProvider

nicolas-grekas reviewed (commented):
Isn't the failure related?

nicolas-grekas reviewed (approved):
(issue fixed while merging)

nicolas-grekas commented:
Thank you @rvanlaak.

rvanlaak commented:
Thank you for moving the backport related fixes to #42918. Apologies for not responding earlier, am back from holidays now 🌴
